tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs

```diff
@@ -0,0 +1,280 @@
+        {
+            Debug.Assert(false, "What is this?");
+            return;
+        }
+
+        _dispatcher.TryEnqueue(() =>
+        {
+            MenuFlyout? flyout = this.ContextFlyout as MenuFlyout;
+            Debug.Assert(flyout is not null, "Why does this button not have a menu flyout?");
+
+            // Update the submenu item for this tool
+            foreach (MenuFlyoutItem menuItem in flyout.Items)
+            {
+                if (menuItem.Tag == tool)
+                {
+                    // Update the name if it's changed
+                    menuItem.Text = tool.Name;
+
+                    var imageIcon = new ImageIcon
+                    {
+                        Source = tool.ToolIcon,
+                    };
+
+                    menuItem.Icon = imageIcon;
+
+                    var menuSubItemFlyout = menuItem.ContextFlyout as MenuFlyout;
+                    Debug.Assert(menuSubItemFlyout != null, "Why is this null?");
+
+                    var pinSubItemItem = menuSubItemFlyout.Items[0] as MenuFlyoutItem;
+                    Debug.Assert(pinSubItemItem != null, "Why is this null?");
+
+                    var unPinSubItemItem = menuSubItemFlyout.Items[1] as MenuFlyoutItem;
+                    Debug.Assert(unPinSubItemItem != null, "Why is this null?");
+
+                    pinSubItemItem.Visibility = tool.IsPinned ? Visibility.Collapsed : Visibility.Visible;
+                    unPinSubItemItem.Visibility = tool.IsPinned ? Visibility.Visible : Visibility.Collapsed;
+                    break;
+                }
+            }
+        });
+    }
+
+    public void ExternalToolMenuItem_Click(object sender, RoutedEventArgs e)
+    {
+        if (sender is MenuFlyoutItem clickedMenuItem)
+        {
+            if (clickedMenuItem.Tag is ExternalTool tool)
+            {
+                ExternalToolLaunchRequest?.Invoke(this, tool);
+
+                // InvokeTool(tool, TargetAppData.Instance.TargetProcess?.Id, TargetAppData.Instance.HWnd);
+            }
+        }
+    }
+
+    private void ExternalToolPinUnpin_Click(object sender, RoutedEventArgs e)
+    {
+        ExternalTool tool = GetToolFromSender(sender);
+        tool.IsPinned = !tool.IsPinned;
+        HideFlyout(sender);
+    }
+
+    public void UnregisterMenuItem_Click(object sender, RoutedEventArgs e)
+    {
+        ExternalTool tool = GetToolFromSender(sender);
+        tool.UnregisterTool();
+
+        HideFlyout(sender);
+    }
+
+    private ExternalTool GetToolFromSender(object sender)
+    {
+        MenuFlyoutItem? clickedMenuItem = sender as MenuFlyoutItem;
+        Debug.Assert(clickedMenuItem != null, "Why is this null?");
+
+        ExternalTool? tool = clickedMenuItem.Tag as ExternalTool;
+        Debug.Assert(tool != null, "Why is this null?");
+
+        return tool;
+    }
+
+    private void HideFlyout(object sender)
+    {
+        if (this.ContextFlyout is MenuFlyout flyout)
```

Comment:
So much of this code casts `ContextFlyout` to `MenuItem` why?  Does the ContextFlyout change what type it is through out the application's lifetime?

Reply:
The cast is not needed in this particular function... so I'll remove it.

But in general, the code is programmatically building a MenuFlyout and manipulating the MenuFlyoutItems based on the state of the external tools.

WinUI3 accepts this MenuFlyout as a FlyoutBase object (stored in ContextFlyout). I can't get MenuFlyoutItems directly from a FlyoutBase, which is why there is the cast to a MenuFlyout.

Does that answer your question?